catalyst-gateway/bin/src/service/api/cardano/cip36/filter.rs

```diff
@@ -0,0 +1,464 @@
+//! Implementation of the GET `/cardano/cip36` endpoint
+
+use std::{cmp::Reverse, sync::Arc};
+
+use futures::StreamExt;
+use poem::http::StatusCode;
+
+use super::{
+    cardano::{
+        cip19_shelley_address::Cip19ShelleyAddress, hash28::HexEncodedHash28, nonce::Nonce,
+        txn_index::TxnIndex,
+    },
+    common::types::generic::error_msg::ErrorMessage,
+    response::{
+        AllRegistration, Cip36Details, Cip36Registration, Cip36RegistrationList,
+        Cip36RegistrationsForVotingPublicKey,
+    },
+    Ed25519HexEncodedPublicKey, SlotNo,
+};
+use crate::db::index::{
+    queries::registrations::{
+        get_all_stakes_and_vote_keys::{
+            GetAllStakesAndVoteKeysParams, GetAllStakesAndVoteKeysQuery,
+        },
+        get_from_stake_addr::{GetRegistrationParams, GetRegistrationQuery},
+        get_from_stake_hash::{GetStakeAddrParams, GetStakeAddrQuery},
+        get_from_vote_key::{GetStakeAddrFromVoteKeyParams, GetStakeAddrFromVoteKeyQuery},
+        get_invalid::{GetInvalidRegistrationParams, GetInvalidRegistrationQuery},
+    },
+    session::CassandraSession,
+};
+
+/// Get registration given a stake key hash, it can be time specific based on asat param,
+/// or the latest registration returned if no asat given.
+pub(crate) async fn get_registration_given_stake_key_hash(
+    stake_hash: HexEncodedHash28, session: Arc<CassandraSession>, asat: Option<SlotNo>,
+) -> AllRegistration {
+    // Get stake addr associated with given stake hash.
+    let mut stake_addr_iter = match GetStakeAddrQuery::execute(
+        &session,
+        GetStakeAddrParams::new(stake_hash.into()),
+    )
+    .await
+    {
+        Ok(stake_addr) => stake_addr,
+        Err(err) => {
+            return AllRegistration::handle_error(&anyhow::anyhow!(
+                "Failed to query stake addr from stake hash {err:?}",
+            ));
+        },
+    };
+
+    if let Some(row_stake_addr) = stake_addr_iter.next().await {
+        let row = match row_stake_addr {
+            Ok(r) => r,
+            Err(err) => {
+                return AllRegistration::handle_error(&anyhow::anyhow!(
+                    "Failed to query stake addr from stake hash {err:?}",
+                ));
+            },
+        };
+
+        // Stake hash successfully converted into associated stake pub key which we use to lookup
+        // registrations.
+        let stake_pub_key = match Ed25519HexEncodedPublicKey::try_from(row.stake_address.clone()) {
+            Ok(key) => key,
+            Err(err) => {
+                return AllRegistration::internal_error(&anyhow::anyhow!(
+                    "Failed to type stake address {err:?}",
+                ));
+            },
+        };
+
+        return get_registration_from_stake_addr(stake_pub_key, asat, session, None).await;
+    }
+    AllRegistration::unprocessable_content(vec![poem::Error::from_string(
```

Comment:
Seems like it should be a `404` Not Found response, like this one
https://github.com/input-output-hk/catalyst-voices/blob/581033186f360510b2e00670190b9487789863a1/catalyst-gateway/bin/src/service/api/cardano/sync_state_get.rs#L19

Reply:
done